core/src/main/scala/chisel3/RawModule.scala

```diff
@@ -296,6 +296,19 @@ abstract class RawModule extends BaseModule {
   protected[chisel3] override def moduleBuilt(): Unit = _afterModuleBuilt.foreach(_())
 
   private[chisel3] def initializeInParent(): Unit = {}
+
+  final var _knownLayers: Option[Seq[layer.Layer]] = None
+  atModuleBodyEnd {
```

Comment:
are there any sneaky APIs that *add* layers with `atModuleBodyEnd`? Can bored probes impact this?

Reply:
Boring could definitely cause this to be incorrect. I can probably show that this is a problem if it's not updated there.

The thing that's really annoying about this is that really _only_ `ExtModule`/`BlackBox` should have known layers. However, because there are APIs that let you create an `Instance` from a `Definition` without examining the `DesignAnnotation` (which has layer information) then this forces everything up to `BaseModule` to need to record this.

It perhaps makes more sense this way, just this is surprisingly invasive just to get the separate elaboration examples working correctly.

@jackkoenig: Do you have any ideas on how to architect this more cleanly?

Reply:
The problem with `Definition` is that it's currently too generic. I.e., you can have a `Definition[A]` for any `A`. Hence, it's weird to have this live on `Definition[A]`, when it's really a property of `A`. (I generally agree with this, though, and we should move towards having a `Definition` be more tied to a module/extmodule and less generic.)